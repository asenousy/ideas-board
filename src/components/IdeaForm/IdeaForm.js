import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import styles from "./IdeaForm.module.css";
import configs from "../../configs";

const textAreaRows = 5;
const textAreaCols = 20;

class IdeaForm extends PureComponent {
  constructor(props) {
    super(props);
    this.state = { ...props.idea };
    this.titleInput = React.createRef();
    this.focusOnTarget = this.focusOnTarget.bind(this);
  }

  // this diversion allows enzyme snapshot testing to work with React.createRef()
  // for more reading https://medium.com/@fabianterh/fixing-typeerror-cannot-read-property-property-of-null-in-react-test-renderer-d91d45137de9
  focusOnTarget() {
    this.titleInput.current.focus();
  }

  componentDidMount() {
    this.focusOnTarget();
  }

  render() {
    const { title, description } = this.state;

    return (
      <form className={styles.wrapper} onSubmit={this.submitHandler}>
        <label className={styles.label}>
          Title:
          <input
            className={styles.input}
            placeholder="title"
            name="title"
            value={title}
            onChange={this.changeHandler}
            ref={this.titleInput}
          />
        </label>
        <label className={styles.label}>
          Description: <br />
          <textarea
            className={styles.textArea}
            placeholder="description"
            name="description"
            rows={textAreaRows}
            cols={textAreaCols}
            value={description}
            onChange={this.changeHandler}
            maxLength={configs.descriptionMaxChars}
          />
        </label>
        <button className={styles.button} type="submit">
          Save
        </button>
      </form>
    );
  }

  changeHandler = e => {
    this.setState({
      [e.target.name]: e.target.value
    });
  };

  submitHandler = e => {
    e.preventDefault();
    this.props.onSubmit && this.props.onSubmit(this.state);
  };
}

IdeaForm.propTypes = {
  idea: PropTypes.shape({
    id: PropTypes.string.isRequired,
    title: PropTypes.string,
    description: PropTypes.string,
    date: PropTypes.number.isRequired,
    new: PropTypes.bool
  }).isRequired,
  onSubmit: PropTypes.func
};

export default IdeaForm;
