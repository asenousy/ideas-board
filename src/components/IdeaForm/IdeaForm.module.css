.wrapper {
  margin: auto;
  display: flex;
  flex-direction: column;
  justify-content: center;
  color: purple;
}

.label {
  margin: 10px;
  text-align: center;
}

.input {
  margin: 5px;
  border-radius: 5px;
  height: 20px;
  border: solid 1px purple;
}

.textArea {
  margin: 5px;
  border-radius: 10px;
  border: solid 1px purple;
}

.button {
  margin: auto;
  padding: 5px 10px;
  font-size: 20px;
  border-radius: 10px;
  color: purple;
  border: solid 1px purple;
}
