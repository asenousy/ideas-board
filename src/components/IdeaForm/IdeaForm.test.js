import React from "react";
import { shallow } from "enzyme";
import IdeaForm from "./IdeaForm";

// this allows enzyme snapshot testing to work with React.createRef()
// for more reading https://medium.com/@fabianterh/fixing-typeerror-cannot-read-property-property-of-null-in-react-test-renderer-d91d45137de9
let focusOnTargetSpy = jest.fn();
jest
  .spyOn(IdeaForm.prototype, "focusOnTarget")
  .mockImplementation(focusOnTargetSpy);

jest.spyOn(Date, "now").mockImplementation(() => 111);

let wrapper;
beforeEach(
  () =>
    (wrapper = shallow(
      <IdeaForm
        idea={{
          id: "1",
          title: "title",
          description: "description",
          date: 111,
          new: true
        }}
      />
    ))
);

it("renders without crashing", () => {
  expect(wrapper).toMatchSnapshot();
});

it("submits new idea updated idea", () => {
  const mockFn = jest.fn();
  wrapper.setProps({ onSubmit: mockFn });
  wrapper
    .find("input")
    .simulate("change", { target: { name: "title", value: "updatedTitle" } });

  expect(wrapper.state("title")).toEqual("updatedTitle");

  wrapper.find("form").simulate("submit", {
    preventDefault: () => {}
  });

  expect(mockFn).toHaveBeenCalledWith({
    id: "1",
    title: "updatedTitle",
    description: "description",
    date: 111,
    new: true
  });
});
