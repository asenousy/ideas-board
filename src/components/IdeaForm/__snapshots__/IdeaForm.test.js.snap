// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <IdeaForm
    idea={
      Object {
        "date": 111,
        "description": "description",
        "id": "1",
        "new": true,
        "title": "title",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <label
          className="label"
        >
          Title:
          <input
            className="input"
            name="title"
            onChange={[Function]}
            placeholder="title"
            value="title"
          />
        </label>,
        <label
          className="label"
        >
          Description: 
          <br />
          <textarea
            className="textArea"
            cols={20}
            maxLength={140}
            name="description"
            onChange={[Function]}
            placeholder="description"
            rows={5}
            value="description"
          />
        </label>,
        <button
          className="button"
          type="submit"
        >
          Save
        </button>,
      ],
      "className": "wrapper",
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Title:",
            <input
              className="input"
              name="title"
              onChange={[Function]}
              placeholder="title"
              value="title"
            />,
          ],
          "className": "label",
        },
        "ref": null,
        "rendered": Array [
          "Title:",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "input",
              "name": "title",
              "onChange": [Function],
              "placeholder": "title",
              "value": "title",
            },
            "ref": Object {
              "current": null,
            },
            "rendered": null,
            "type": "input",
          },
        ],
        "type": "label",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Description: ",
            <br />,
            <textarea
              className="textArea"
              cols={20}
              maxLength={140}
              name="description"
              onChange={[Function]}
              placeholder="description"
              rows={5}
              value="description"
            />,
          ],
          "className": "label",
        },
        "ref": null,
        "rendered": Array [
          "Description: ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "br",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "textArea",
              "cols": 20,
              "maxLength": 140,
              "name": "description",
              "onChange": [Function],
              "placeholder": "description",
              "rows": 5,
              "value": "description",
            },
            "ref": null,
            "rendered": null,
            "type": "textarea",
          },
        ],
        "type": "label",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Save",
          "className": "button",
          "type": "submit",
        },
        "ref": null,
        "rendered": "Save",
        "type": "button",
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <label
            className="label"
          >
            Title:
            <input
              className="input"
              name="title"
              onChange={[Function]}
              placeholder="title"
              value="title"
            />
          </label>,
          <label
            className="label"
          >
            Description: 
            <br />
            <textarea
              className="textArea"
              cols={20}
              maxLength={140}
              name="description"
              onChange={[Function]}
              placeholder="description"
              rows={5}
              value="description"
            />
          </label>,
          <button
            className="button"
            type="submit"
          >
            Save
          </button>,
        ],
        "className": "wrapper",
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Title:",
              <input
                className="input"
                name="title"
                onChange={[Function]}
                placeholder="title"
                value="title"
              />,
            ],
            "className": "label",
          },
          "ref": null,
          "rendered": Array [
            "Title:",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "input",
                "name": "title",
                "onChange": [Function],
                "placeholder": "title",
                "value": "title",
              },
              "ref": Object {
                "current": null,
              },
              "rendered": null,
              "type": "input",
            },
          ],
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Description: ",
              <br />,
              <textarea
                className="textArea"
                cols={20}
                maxLength={140}
                name="description"
                onChange={[Function]}
                placeholder="description"
                rows={5}
                value="description"
              />,
            ],
            "className": "label",
          },
          "ref": null,
          "rendered": Array [
            "Description: ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": "br",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "textArea",
                "cols": 20,
                "maxLength": 140,
                "name": "description",
                "onChange": [Function],
                "placeholder": "description",
                "rows": 5,
                "value": "description",
              },
              "ref": null,
              "rendered": null,
              "type": "textarea",
            },
          ],
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Save",
            "className": "button",
            "type": "submit",
          },
          "ref": null,
          "rendered": "Save",
          "type": "button",
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
